fix(api): Fix path planning after failed tip drop

SyntaxColoring opened the pull request:
## Overview

Closes EXEC-764.

## Test Plan and Hands on Testing

* [x] Run any protocol on an OT-2 and make sure tip pickups, tip drops, aspirates, and dispenses still work.
* [x] Run any protocol on a Flex and make sure tip pickups, tip drops, aspirates, and dispenses still work.
* [x] Induce a tip drop error on a Flex and test the error recovery flows, which should already be implemented on the frontend.

  Retrying the tip drop should now work as you'd intuitively expect. In particular, it should *not* cause the pipette to ram the tip into the bottom of the tip rack.

  "Ignore and continue" will probably not work—that's a separate PR. See EXEC-676.

  "Cancel run" should still work, with no change from before.

## Changelog

`opentrons.protocol_engine` and `opentrons.hardware_control` both keep track of tip state. For error recovery purposes, `opentrons.protocol_engine` updates its tip state very specifically and carefully: the tip is logically removed *only if the sensor confirms that it has been physically removed.* But `opentrons.hardware_control` updates its logical state more eagerly than that. The mismatch causes path planning bugs.

The solution is to let `opentrons.protocol_engine` opt out of `opentrons.hardware_control`'s state management and do it itself. In other words, split up the hardware API's `pick_up_tip()` method into one method that just does the physical motion, and several other methods that just do the logical state updates. https://github.com/Opentrons/opentrons/pull/15275 previously did this for `pick_up_tip()`, for the same reasons.

## Review requests

Given the risk here, I'm doing my best to avoid introducing subtle unintentional changes, but there are some.

* I do not understand this part of `OT3API.drop_tip()`:

  ```python
  # call this in case we're simulating:
  if isinstance(self._backend, OT3Simulator):
      self._backend._update_tip_state(realmount, False)
  ```

  Given that it's a state update, it seems like it should **not** live in `tip_drop_moves()`. But *`tip_pickup_moves()`* has the same thing, so I've left it in `tip_drop_moves()` to handle it the same way. Is this correct?

* ~~Is it a problem that when Protocol Engine does a tip drop, it no longer sets the hardware API's current volume to 0?~~
  * It seems like it's not a problem in practice, because we still reset the volume to 0 when we do a tip *pickup.*

## Risk assessment

High. This affects core OT-2 and Flex behavior, and we don't have any automated tests that can be helpful here.

SyntaxColoring committed:
Add drop_tip_moves().

The new drop_tip_moves() function is just the motion part of a drop_tip(), excluding any updates that the hardware API would otherwise make to its own logical state. It is like the existing tip_pickup_moves() function.

Reimplement drop_tip() in terms of drop_tip_moves().

This is just a refactor so far, with no intended behavioral change.

SyntaxColoring committed:
Leave todo comments for cache_tip(), add_tip(), and remove_tip().

SyntaxColoring committed:
Bring `API`'s logical tip removal code out of plan_check_drop_tip().

This is a trivial refactor and should have no behavioral effect. This is to bring make `API.drop_tip()` look structurally the same as `OT3API.drop_tip()`.

SyntaxColoring committed:
Implement drop_tip()'s state updates in terms of other public methods, where possible.

This is not intended to have any behavioral change.

The goal of this is to make it easier for callers (i.e. Protocol Engine) to see what they have to do in order to replicate the total effect of drop_tip(). And to make it easier for human code reviewers to confirm that that's done correctly.

SyntaxColoring committed:
Update Protocol Engine to only change hardware state only if the tip drop succeeds.

This is, finally, the part that makes the actual behavioral change and fixes the actual bug.

Having split up drop_tip(), we can now do the hardware API's state updates only after we've verified that the tip drop has physically  succeeded. This mirrors what Protocol Engine does for its own state after tip drop errors--see TipPhysicallyAttachedError.

SyntaxColoring committed:
Move simulation-specific thing to tip_drop_moves().

I don't understand what this does, and given that it's a state update, it seems like it should not live in tip_drop_moves(). But tip_pickup_moves() has the same thing, so I presume we should treat it the same way.

sfoster1 commented on api/src/opentrons/hardware_control/ot3api.py:
```diff
@@ -2326,11 +2319,27 @@ def _remove_tips() -> None:
 
-        _remove_tips()
-        # call this in case we're simulating
+        # call this in case we're simulating:
```
this simulator is (theoretically) used for more than just analysis. we update this here because the backend is actually simulating the sensor, so we need to inform the simulator that it should now act as if a type is not physically present.

sfoster1 commented on api/src/opentrons/protocol_engine/execution/tip_handler.py:
```diff
@@ -248,6 +249,9 @@ async def pick_up_tip(
         )
 
+        # todo(mm, 2024-10-15): The hardware API's original pick_up_tip() implementation
+        # seems to set the *current* volume to 0, not the working volume.
```
the "working volume" is the amount of volume we can hold - it's basically the smaller of the volume of the tip and the volume of the pipette (you could put a 20ul tip on a 10ul pipette, I believe).

sfoster1 commented on api/src/opentrons/hardware_control/protocols/__init__.py:
```diff
@@ -87,6 +91,10 @@ class FlexHardwareControlInterface(
 
+    # todo(mm, 2024-10-17): This probably belongs in InstrumentConfigurer, alongside
+    # add_tip() and remove_tip().
+    # todo(mm, 2024-10-17): What is the difference between this and add_tip()?
```
`add_tip` will warn and not reconfigure tip details if a tip is already attached to the instrument (this is in `pipette_handler.add_tip`. `cache_tip` overwrites whatever is there and doesn't care whether a tip is already attached. if we removed one, it would be `add_tip`, but I'm not sure whether that's safe or if it's used by e.g. older PAPI versions.

sfoster1 reviewed (commented):
Answered some of your questions. Looks like the right path.

SyntaxColoring committed:
Clean up pick_up_tip() test.

- Consistently use Decoy.
- Remove unused rehearsals.
- Add missing rehearsals.

SyntaxColoring committed:
Update test.

SyntaxColoring committed:
Delete todo about working volume.

sfoster1 reviewed (approved):
niuce, thank you!

SyntaxColoring committed:
Merge branch 'edge' into drop_tip_moves

SyntaxColoring committed:
Update todo comments from answers on PR.

SyntaxColoring committed:
Merge branch 'edge' into drop_tip_moves
